taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -0,0 +1,1022 @@
+	field_amount_delegated    = []byte{6}
+)
+
+var Big10000 = new(big.Int).SetInt64(10000)
+
+// Maximum number of validators per batch returned by getValidators call
+const GetValidatorsMaxCount = 50
+
+// Maximum number of validators per batch returned by getDelegatorDelegations call
+const GetDelegatorDelegationsMaxCount = 50
+
+type State struct {
+	RwardsPer1Stake *big.Int
+
+	// number of references
+	Count uint32
+}
+
+type Contract struct {
+	cfg            Config
+	storage        StorageWrapper
+	delayedStorage Reader
+	Abi            abi.ABI
+
+	validators    Validators
+	delegations   Delegations
+	undelegations Undelegations
+
+	eligible_count_orig      uint64
+	eligible_count           uint64
+	eligible_vote_count_orig uint64
+	eligible_vote_count      uint64
+	amount_delegated_orig    *big.Int
+	amount_delegated         *big.Int
+	lazy_init_done           bool
+}
+
+func (self *Contract) Init(cfg Config, storage Storage, readStorage Reader) *Contract {
+	self.cfg = cfg
+	self.storage.Init(storage)
+	self.delayedStorage = readStorage
+	self.Abi, _ = abi.JSON(strings.NewReader(TaraxaDposClientMetaData))
+
+	self.validators.Init(&self.storage, field_validators)
+	self.delegations.Init(&self.storage, field_delegations)
+	self.undelegations.Init(&self.storage, field_undelegations)
+
+	return self
+}
+
+func (self *Contract) UpdateStorage(readStorage Reader) {
+	self.delayedStorage = readStorage
+}
+
+///// DO WE NEED THIS ? ////
+func (self *Contract) UpdateConfig(cfg Config) {
+	self.cfg = cfg
+}
+
+////////////////////////////////////////////////////////////////
+
+func (self *Contract) Register(registry func(*common.Address, vm.PrecompiledContract)) {
+	defensive_copy := *contract_address
+	registry(&defensive_copy, self)
+}
+
+func (self *Contract) RequiredGas(ctx vm.CallFrame, evm *vm.EVM) uint64 {
+	// TODO: based on method being called, calculate the gas somehow. Doing it based on the length of input is totally useless
+	return uint64(len(ctx.Input)) * 20
+}
+
+func (self *Contract) lazy_init() {
+	if self.lazy_init_done {
+		return
+	}
+	self.lazy_init_done = true
+	self.storage.Get(stor_k_1(field_eligible_count), func(bytes []byte) {
+		self.eligible_count_orig = bin.DEC_b_endian_compact_64(bytes)
+	})
+	self.eligible_count = self.eligible_count_orig
+	self.storage.Get(stor_k_1(field_eligible_vote_count), func(bytes []byte) {
+		self.eligible_vote_count_orig = bin.DEC_b_endian_compact_64(bytes)
+	})
+	self.eligible_vote_count = self.eligible_vote_count_orig
+	self.amount_delegated_orig = bigutil.Big0
+	self.storage.Get(stor_k_1(field_amount_delegated), func(bytes []byte) {
+		self.amount_delegated_orig = bigutil.FromBytes(bytes)
+	})
+	self.amount_delegated = self.amount_delegated_orig
+}
+
+func (self *Contract) BeginBlockCall(rewards map[common.Address]*big.Int) {
+	for validator, reward := range rewards {
+		self.update_rewards(&validator, reward)
+	}
+}
+
+func (self *Contract) EndBlockCall() {
```

Comment:
I think all of these values can and should be adjusted inside functions like delegate/undelegate/etc... and not here. If user delegates and that means we need to adjust amount_delegated in database, it should be done in his tx and he should also pay for it...

Reply:
it does not matter who will pay for it, this is optimization to save it only one per block if there is a change. It should be here

Reply:
in this case no one pays for it. It is basically free storage change

Reply:
what? you will pay it in *delegate transaction. Anyways we do not have any fee architecture yet